pump/server_test.go

```diff
@@ -555,3 +562,272 @@ func (s *listenSuite) TestReturnListener(c *C) {
+	origUtilGetTSO func(pd.Client) (int64, error)
+}
+
+var _ = Suite(&startServerSuite{})
+
+func (s *startServerSuite) SetUpTest(c *C) {
+	s.origUtilGetTSO = utilGetTSO
+}
+
+func (s *startServerSuite) TearDownTest(c *C) {
+	utilGetTSO = s.origUtilGetTSO
+}
+
+func (s *startServerSuite) TestStartPumpServer(c *C) {
+	utilGetTSO = func(pd.Client) (int64, error) {
+		return 0, nil
+	}
+	etcdClient := testEtcdCluster.RandClient()
+	cfg := &Config{
+		ListenAddr:        "http://127.0.0.1:8250",
+		AdvertiseAddr:     "http://127.0.0.1:8260",
+		Socket:            "unix://127.0.0.1:" + string(time.Now().UnixNano()) + "/hello/world",
+		EtcdURLs:          strings.Join(etcdClient.Endpoints(), ","),
+		DataDir:           "/tmp/pump",
+		HeartbeatInterval: 1500,
+		LogLevel:          "debug",
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	grpcOpts := []grpc.ServerOption{grpc.MaxRecvMsgSize(GlobalConfig.maxMsgSize)}
+	sig := make(chan struct{})
+	startNodeImpl := &startNode{status: &node.Status{State: node.Online, NodeID: "startnode-long", Addr: "http://192.168.199.100:8260"}}
+	p := &Server{
+		dataDir:    cfg.DataDir,
+		storage:    &startStorage{sig: sig},
+		clusterID:  8012,
+		node:       startNodeImpl,
+		tcpAddr:    cfg.ListenAddr,
+		unixAddr:   cfg.Socket,
+		gs:         grpc.NewServer(grpcOpts...),
+		ctx:        ctx,
+		cancel:     cancel,
+		tiStore:    nil,
+		gcDuration: time.Duration(cfg.GC) * 24 * time.Hour,
+		pdCli:      nil,
+		cfg:        cfg,
+		triggerGC:  make(chan time.Time)}
+	defer p.Close()
+	go p.Start()
+
+	// wait until the server is online
+	timeEnd := time.After(5 * time.Second)
+	getInterval := time.NewTicker(500 * time.Microsecond)
+WAIT:
+	for {
+		select {
+		case <-getInterval.C:
+			resp, err := http.Get("http://127.0.0.1:8250/status")
+			if err != nil {
+				if !strings.Contains(err.Error(), "connect: connection refused") {
+					c.Fatal("Connect to pump server fail, error is " + err.Error())
+				}
+				continue
+			}
+			// should receive valid and correct node status info
+			c.Assert(resp, NotNil)
+			defer resp.Body.Close()
+			bodyByte, err := ioutil.ReadAll(resp.Body)
+			c.Assert(err, IsNil)
+
+			nodesStatus := &HTTPStatus{}
+			err = json.Unmarshal(bodyByte, nodesStatus)
+			c.Assert(err, IsNil)
+			c.Assert(nodesStatus.ErrMsg, Equals, "")
+			c.Assert(nodesStatus.CommitTS, Equals, int64(0))
+			fakeNodeStatus, ok := nodesStatus.StatusMap["startnode-long"]
+			c.Assert(ok, IsTrue)
+			c.Assert(fakeNodeStatus.NodeID, Equals, "startnode-long")
+			c.Assert(fakeNodeStatus.Addr, Equals, "http://192.168.199.100:8260")
+			c.Assert(fakeNodeStatus.State, Equals, node.Online)
+			break WAIT
+		case <-timeEnd:
+			getInterval.Stop()
+			c.Fatal("Wait pump to be online for more than 5 seconds")
+		}
+	}
+	getInterval.Stop()
+
+	// test AllDrainer
+	resultStr := httpRequest(c, http.MethodGet, "http://127.0.0.1:8250/drainers")
+	c.Assert(strings.Contains(resultStr, "can't provide service"), IsTrue)
```

Comment:
Have you tried `c.Assert(resultStr, Matches, xxx)`?

Reply:
Revised

Reply:
Revised